python/src/iceberg/table/metadata.py

```diff
@@ -0,0 +1,366 @@
+
+    default_sort_order_id: int
+    """Default sort order id of the table. Note that this could be used by 
+    writers, but is not used when reading because reads use the specs stored
+     in manifest files."""
+
+    def validate(self) -> None:
+        """Checks that the table metadata object is valid. The validation schema
+        used depends on the Iceberg table metadata version."""
+        if self.format_version == 1:
+            self.validate_v1(self.to_dict())
+        elif self.format_version == 2:
+            self.validate_v2(self.to_dict())
+        else:
+            raise ValueError(f"Unknown table metadata version {self.format_version}")
+
+    @staticmethod
+    def validate_v1(metadata: dict) -> None:
+        """Perform a JSONSchema validation using the v1 Iceberg table metadata schema"""
+        try:
+            validate_json(instance=metadata, schema=TABLE_METADATA_V1_SCHEMA)
+        except ValidationError as e:
+            # TODO Log something here
+            raise (e)
+
+    @staticmethod
+    def validate_v2(metadata: dict) -> None:
+        """Perform a JSONSchema validation using the v2 Iceberg table metadata schema"""
+        try:
+            validate_json(instance=metadata, schema=TABLE_METADATA_V2_SCHEMA)
+        except ValidationError as e:
+            # TODO Log something here
+            raise (e)
+
+    @classmethod
+    def from_byte_stream(cls, byte_stream, encoding: str = "utf-8") -> "TableMetadata":
+        """Instantiate a TableMetadata object from a byte stream
+
+        Args:
+            byte_stream: A file-like byte stream object
+            encoding (default "utf-8"): The byte encoder to use for the reader
+        """
+        reader = codecs.getreader(encoding)
+        metadata = json.load(reader(byte_stream))
+        return cls.from_dict(metadata)
+
+    @classmethod
+    def from_input_file(cls, input_file: InputFile, encoding: str = "utf-8") -> "TableMetadata":
+        """Create a TableMetadata instance from an input file
+
+        Args:
+            input_file (InputFile): A custom implementation of the iceberg.io.file.InputFile abstract
+            base class
+            encoding (str): Encoding to use when loading bytestream
+
+        Returns:
+            TableMetadata: A table metadata instance
+
+        """
+        return cls.from_byte_stream(byte_stream=input_file.open(), encoding=encoding)
+
+    def to_output_file(self, output_file: OutputFile, overwrite: bool = False) -> None:
+        """Write a TableMetadata instance to an output file
+
+        Args:
+            output_file (OutputFile): A custom implementation of the iceberg.io.file.OutputFile abstract
+            base class
+            overwrite (bool): Where to overwrite the file if it already exists. Defaults to `False`.
+        """
+        f = output_file.create(overwrite=overwrite)
+        f.write(json.dumps(self.to_dict()).encode("utf-8"))
+
```

Comment:
Should we create a helper [Context Manger](https://docs.python.org/3/library/stdtypes.html#typecontextmanager) like `TableMetaDataWriter` for this?

```python
class output_file_writer():
    def __init__(self, output_file: OutputFile, overwrite: bool = False):
        self.__output_file = output_file
        self.__overwrite = True if overwrite else False

    def __enter__(self):
        self.__file = output_file.create(self.__overwrite)
        return self

    def __exit__(self, type, val, tb):
        self.__file.close()
```

With that helper the above method becomes:
```python
def to_output_file(self, output_file: OutputFile, overwrite: bool = False) -> None:
        with output_file_writer(output_file, overwrite) as file:
            file.write(json.dumps(self.to_dict()).encode("utf-8"))
```
You could even default the encoding to UTF-8 inside the context manager, in one single place, if that is our standard encoding format.

Reply:
I'm a big fan of context managers, but I think they should be integrated into the class itself, instead of using yet another wrapper.

I'm quite confused btw `output_file.create(overwrite=overwrite)` should return an `OutputStream` from `iceberg.io.base`, but that class doesn't seem to exist 🤔

Reply:
[OutputStream](https://github.com/apache/iceberg/blob/master/python/src/iceberg/io/base.py#L61) is just a protocol, so it's basically documentation that explains [OutputFile.create(...)](https://github.com/apache/iceberg/blob/master/python/src/iceberg/io/base.py#L160) should return an object with a `write`, `closed` and `close` method.